packages/eservice-template-instances-updater/src/eserviceTemplateInstancesUpdaterConsumerServiceV2.ts

```diff
@@ -0,0 +1,430 @@
+  }
+): EServiceTemplateVersionV2 {
+  const eserviceTemplate = getTemplateFromEvent(msg);
+
+  const eserviceTemplateVersion = eserviceTemplate.versions.find(
+    (v) => v.id === msg.data.eserviceTemplateVersionId
+  );
+
+  if (!eserviceTemplateVersion) {
+    throw missingKafkaMessageDataError("eserviceTemplateVersion", msg.type);
+  }
+
+  return eserviceTemplateVersion;
+}
+
+async function commitUpdateToInstanceDescriptors(
+  msg: EServiceTemplateEventEnvelope & {
+    data: { eserviceTemplateVersionId: string };
+  },
+  refreshableToken: RefreshableInteropToken,
+  correlationId: CorrelationId,
+  action: (
+    eservice: catalogApi.EService,
+    descriptor: catalogApi.EServiceDescriptor,
+    headers: InteropHeaders
+  ) => Promise<void>
+): Promise<void> {
+  await commitUpdateToInstances(
+    msg,
+    refreshableToken,
+    correlationId,
+    async (instance, headers) => {
+      const instanceDescriptor = instance.descriptors.find(
+        (d) => d.templateVersionRef?.id === msg.data.eserviceTemplateVersionId
+      );
+
+      if (instanceDescriptor) {
+        await action(instance, instanceDescriptor, headers);
+      }
+    }
+  );
+}
+
+async function commitUpdateToInstances(
+  msg: EServiceTemplateEventEnvelope,
+  refreshableToken: RefreshableInteropToken,
+  correlationId: CorrelationId,
+  action: (
+    eservice: catalogApi.EService,
+    headers: InteropHeaders
+  ) => Promise<void>
+): Promise<void> {
+  const token = (await refreshableToken.get()).serialized;
+  const headers = getInteropHeaders({
+    token,
+    correlationId,
+  });
+
+  const instances = await getAllEServiceTemplateInstances(
+    getTemplateFromEvent(msg).id,
+    headers
+  );
+
+  const chunkSize = 10;
+  // Call the action on each instance in chunks
+  // eslint-disable-next-line functional/no-let
+  for (let i = 0; i < instances.length; i += chunkSize) {
+    const chunk = instances.slice(i, i + chunkSize);
+    await Promise.all(chunk.map((instance) => action(instance, headers)));
+  }
+}
+
+async function cloneDocument(
+  doc: catalogApi.EServiceDoc,
+  fileManager: FileManager,
+  logger: Logger
+): Promise<catalogApi.CreateEServiceDescriptorDocumentSeed> {
+  const clonedDocumentId = randomUUID();
+
+  const clonedPath = await fileManager.copy(
+    config.eserviceDocumentsContainer,
+    doc.path,
+    config.eserviceDocumentsPath,
+    clonedDocumentId,
+    doc.name,
+    logger
+  );
+
+  return {
+    documentId: clonedDocumentId,
+    kind: "DOCUMENT",
+    contentType: doc.contentType,
+    prettyName: doc.prettyName,
+    fileName: doc.name,
+    filePath: clonedPath,
+    checksum: doc.checksum,
+    serverUrls: [],
+  };
+}
+
+async function getAllEServiceTemplateInstances(
+  eserviceTemplateId: string,
+  headers: InteropHeaders
+): Promise<catalogApi.EService[]> {
+  return await getAllFromPaginated<catalogApi.EService>(
+    async (offset, limit) =>
+      await catalogProcess.client.getEServices({
```

Comment:
Discussed offline, leaving a comment here to keep track 👇 
All the other consumers that react to events and perform side effects read data from the readmodel, not from API calls (see, for example, `delegation-items-archiver` or `eservice-descriptors-archiver`). 

This call would probably fail because it does not accept internal as a role, and it's not meant to be called with an internal token because it performs checks on the tenant/organizationId (not present in internal tokens) to restrict eservice visibility.

I would adopt the same approach used by other consumers

Reply:
Great suggestion! Implemented in this [commit](https://github.com/pagopa/interop-be-monorepo/pull/1487/commits/1951896072365df8518b5401306beec4e07ae7e9)